components/sup/src/ctl_gateway/server.rs

```diff
@@ -481,14 +481,20 @@ pub fn run(listen_addr: SocketAddr, secret_key: String, mgr_sender: MgrSender) {
           let server =
               TcpListener::bind(&listen_addr).expect("Could not bind ctl gateway listen address!")
                                              .incoming()
-                                             .map(|tcp_stream| {
-                                                 let addr =
-                                                     tcp_stream.peer_addr().expect("Couldn't get \
-                                                                                    peer address!");
-                                                 let io = SrvCodec::new().framed(tcp_stream);
-                                                 let client = Client { handle: handle.clone(),
-                                                                       state:  state.clone(), };
-                                                 (client.serve(io), addr)
+                                             // Filter out instances where there is no peer address
+                                             // on the socket. peer_addr() can be Err when a client
+                                             // disregards the protocol such as from a `netcat -z`
+                                             // connection.
+                                             .filter_map(|tcp_stream| {
+                                                 if tcp_stream.peer_addr().is_ok() {
+                                                     let addr = tcp_stream.peer_addr().unwrap();
```

Comment:
There might be something obvious I am missing here but doesn't this `unwrap` create the same panic risk as the `expect`?

Reply:
No, because on the line above, we've explicitly checked to make sure it's `Ok`

Reply:
oh of course! I was missing something obvious.

Reply:
The line right above with `if tcp_stream.peer_addr().is_ok() {` acts as a guard to prevent the `.unwrap()` on an `Err` from ever occuring.

However, this [commit](https://github.com/habitat-sh/habitat/pull/7244/commits/7f8f72980c20c182d2cceac29a397de659c4f98c) has an improvement that avoids `unwrap()` altogether.